engine/proxy.go

```diff
@@ -0,0 +1,227 @@
+	"github.com/prisma/prisma-client-go/logger"
+	"github.com/prisma/prisma-client-go/runtime/types"
+)
+
+func NewDataProxyEngine(schema, schemaEnvVarName string) *DataProxyEngine {
+	return &DataProxyEngine{
+		Schema:           schema,
+		schemaEnvVarName: schemaEnvVarName,
+		http:             &http.Client{},
+	}
+}
+
+type DataProxyEngine struct {
+	// http is the internal http client
+	http *http.Client
+
+	// url holds the query-engine url
+	url string
+
+	// schemaEnvVarName is the env var for the datasource url
+	// this is needed internally to extract the api key
+	schemaEnvVarName string
+
+	// Schema contains the prisma Schema
+	Schema string
+
+	// apiKey contains the parsed prisma data proxy api key from the connection string
+	apiKey string
+}
+
+func (e *DataProxyEngine) Connect() error {
+	// Example uri: https://aws-eu-west-1.prisma-data.com/2.26.0/412bf0a1742a576d699fbd5102a4f725557eff3992995f2e18febce128794961/
+	hash := hashSchema(e.Schema)
+	logger.Debug.Printf("local schema hash %s", hash)
+
+	logger.Debug.Printf("parsing connection string from database url %s", e.schemaEnvVarName)
+
+	connectionString := os.Getenv(e.schemaEnvVarName)
+	if connectionString == "" {
+		return fmt.Errorf("no connection string found")
+	}
+	u, err := url.Parse(connectionString)
+	if err != nil {
+		return fmt.Errorf("parse prisma string: %w", err)
+	}
+	e.apiKey = u.Query().Get("api_key")
+	if e.apiKey == "" {
+		return fmt.Errorf("could not parse api key from data proxy prisma connection string")
+	}
+
+	e.url = getCloudURI(u.Host, hash)
+	// TODO temp: remove this once prisma data proxy works with recent stable versions – right now only 3.0.1 works
+	e.url = "https://" + path.Join(u.Host, "3.0.1", hash)
+	logger.Debug.Printf("using %s as remote URI", e.url)
+	if err := e.uploadSchema(context.Background()); err != nil {
+		return fmt.Errorf("upload schema: %w", err)
+	}
+
+	return nil
+}
+
+func (e *DataProxyEngine) uploadSchema(ctx context.Context) error {
+	logger.Debug.Printf("uploading schema...")
+	b64Schema := encodeSchema(e.Schema)
+	res, err := e.request(ctx, "PUT", "/schema", []byte(b64Schema))
+	if err != nil {
+		return fmt.Errorf("put schema: %w", err)
+	}
+	type SchemaResponse struct {
+		SchemaHash string `json:"schemaHash"`
+	}
+	var response SchemaResponse
+	if err := json.Unmarshal(res, &response); err != nil {
+		return fmt.Errorf("schema response err: %w", err)
+	}
+	logger.Debug.Printf("remote schema hash %s", response.SchemaHash)
+	logger.Debug.Printf("schema upload done.")
+	return nil
+}
+
+func (e *DataProxyEngine) Disconnect() error {
+	return nil
+}
+
+func (e *DataProxyEngine) Do(ctx context.Context, payload interface{}, into interface{}) error {
+	startReq := time.Now()
+	data, err := json.Marshal(payload)
+	if err != nil {
+		return fmt.Errorf("payload marshal: %w", err)
+	}
+
+	body, err := e.request(ctx, "POST", "/graphql", data)
+	if err != nil {
+		return fmt.Errorf("request failed: %w", err)
```

Comment:
If this doesn't try to upload a schema on a SchemaMissing 404 error _somewhere_ when handling this error, then we'll have a pretty fragile client.